### ISSUE TITLE
Reproduced the simulation experiment in pytorch [Converges but different results for GIoU from the paper] 

### ISSUE THREAD

**Issue by ksachdeva** [Zzh-tju/DIoU#22]
Hi,

First thanks for the very interesting work. 

I tried to reproduce the results of your simulation experiment and faced following challenges -

a) Tried to run it using Octave (as I do not have matlab) and it was quite slow ... as in many days to finish it. Most likely octave issues!

b) Reproduced the simulation algorithm in pytorch but it required the loss to be scalar to compute the gradients. However, your experiment computes gradient per sample. Fortunately using the newly introduced vmap API in pytorch I have managed to do that as well. 

c) On my mac M1, every loss function experiment takes less than a minute. I am providing the implementation in Google colab where it takes from 1 to 3 minutes per loss function. 

d) Finally, my code assumes xyxy format for boxes, loss and gradient update. As such (based on my understanding of your matlab scripts) you also compute gradients in xyxy space and then translate them to cxycywh before updating the anchor boxes. I would like to think that this is not an issue but please confirm if you think it is the contributing factor that creates differences between your and my results.

Your experiment [from the paper]
![image](https://github.com/Zzh-tju/DIoU/assets/127006/c8b42ddf-3b0d-40df-aea9-b56db29b0ec0)

My implementation
![image](https://github.com/Zzh-tju/DIoU/assets/127006/34bbc313-0680-43cf-b8fe-8dd225a11f5b)

As per my experiment code, the giou loss reduces faster than others.

Most likely, I am not doing something correct and would appreciate if you could look at the notebook and point out why my results are different from yours. 

Here is the link to the google colab notebook which has everything that is needed to run the experiment. 
https://colab.research.google.com/drive/1GAXn6tbd7rKZ1iuUK1pIom_R9rTH1eVU?usp=sharing

Many thanks for the wonderful paper and work you have done in this field.

Regards
Kapil

**Comment by Zzh-tju** [comment]
This phenomenon is likely due to the difference between the derivative of pytorch and the customized derivative. The gradient calculation in matlab version is consistent with the C version used in DarkNet-YOLOv3. For the box edges, there are some special overlapping cases where the direction of box optimization will be  geometrically opposite if  strict mathematical derivatives are followed. Therefore, by manually defining some directions for those cases, the gradient descent will always face the direction that is conducive to minimizing geometric factor differences.

**Comment by ksachdeva** [comment]
Thanks @Zzh-tju for looking into this and the insight you gave.

I am assuming that you have looked into my notebook and have not seen any silly mistake I may be making. 

I would think that in a more practical experiment i.e. using a dataset like COCO and VOC, the superiority of CIOU would show up. I will try that as well and will report back.

Thanks again.

**Comment by ksachdeva** [comment]
Hi @Zzh-tju 

I did the experiment using yolov5 and CIoU performs better (in terms of converging faster). 

Here is the repo with the details and results of the experiment.
https://github.com/ksachdeva/od-loss-exp

I also made a tutorial to explain your work
https://www.youtube.com/watch?v=4wXXNQ4Ylrk

Thanks

Regards
Kapil
